src/gui/lazyfilelist.h

```diff
@@ -0,0 +1,329 @@
+            // special case code - add a ".."
+            files[0].SetDirUp();
+            --windowStartingFrom;
+            return true;
+        }
+
+        F_DIR_RAII_Iterator dir(path);
+        // clear the item at the zeroth position
+        files[0].Clear();
+        while (dir.FindNext()) {
+            if (LessEF(files[0], dir.fno) && LessFE(dir.fno, files[1])) {
+                // to be inserted, the entry must be greater than zeroth entry AND less than the first entry
+                files[0].CopyFrom(dir.fno);
+            }
+        }
+        --windowStartingFrom;
+        return true;
+    }
+
+    /// A similar operation like MoveUp, but moves the window down by 1 item.
+    /// @return true if the window was actually moved
+    bool MoveDown() {
+        if (totalFiles <= windowStartingFrom + WINDOW_SIZE + 1) {
+            return false; // no more files
+        }
+        std::rotate(files.begin(), files.begin() + 1, files.end());
+
+        F_DIR_RAII_Iterator dir(path);
+        // prepare the last entry for comparison in the do-while cycle
+        files[WINDOW_SIZE - 1].MakeLastFName();
+        while (dir.FindNext()) {
+            if (LessFE(dir.fno, files[WINDOW_SIZE - 1]) && LessEF(files[WINDOW_SIZE - 2], dir.fno)) {
+                // to be inserted, the entry must be greater than the pre-last entry AND less than the last entry
+                files[WINDOW_SIZE - 1].CopyFrom(dir.fno);
+            }
+        }
+        ++windowStartingFrom;
+        return true;
+    }
+
+private:
+    struct Entry {
+        bool isFile;
+        char name[96];
+        uint16_t date, time;
+        void Clear() {
+            isFile = false;
+            name[0] = 0;
+            date = time = 0;
+        }
+        void CopyFrom(const FILINFO &fno) {
+            strncpy(name, fno.fname, 96);
+            isFile = (fno.fattrib & AM_DIR) == 0;
+            date = fno.fdate;
+            time = fno.ftime;
+        }
+        void SetDirUp() {
+            name[0] = name[1] = '.';
+            name[2] = 0;
+            isFile = false;
+            date = time = USHRT_MAX;
+        }
+        void MakeLastFName() {
+            isFile = true;
+            name[0] = 0;
+            date = time = 0;
+        }
+    };
+
+    std::array<Entry, WINDOW_SIZE> files; ///< roughly WINDOW_SIZE * 100B, can be placed in CCRAM if needed. For MINI it is only 9 entries -> only 900B
+    size_t filesInWindow;                 ///< number of files populated in the window - less than WINDOW_SIZE when there are less files in the dir
+    size_t totalFiles;                    ///< total number of entries in the directory
+    int windowStartingFrom;               ///< from which entry index the window starts (e.g. from the 3rd file in dir).
+                                          ///< intentionally int, because -1 means ".."
+    char path[256];                       ///< current directory path - @@TODO this may not be enough - needs checking
```

Comment:
We already have `F_MAXPATHNAMELENGTH`... (currently defined in `window_file_list.h`). Maybe would be nice to make sure all those things are using the same value originating from one place.

Reply:
I don't want to pollute lazyfilelist.h with window_file_list.h or extracting the F_MAXPATHNAMELENGTH into a separate header for now, this will get sorted out when window_file_list gets upgraded to C++.